File: Leetcode/JS/Leetcode629-K个逆序对.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number}
 */
 var kInversePairs = function(n, k) {
    let f = new Array(n + 1).fill(0).map(() => new Array(k + 1).fill(0));
    f[1][0] = 1;
    let mod = 1e9 + 7;
    for(let i = 2;i <= n;i++){
        let s = 0;
        for(let j = 0;j <= k;j++){
            s += f[i - 1][j];
            if(j - i >= 0) s -= f[i - 1][j - i];
            f[i][j] = s % mod;
        }
    }
    return f[n][k];
};

File: Leetcode/sbgo/Leetcode629-K个逆序对.go
package sbgo

func kInversePairs(n int, k int) int {
	f := make([][]int, n+1)
	for i := range f {
		f[i] = make([]int, k+1)
	}
	f[1][0] = 1
	const mod = 1e9 + 7
	for i := 2; i <= n; i++ {
		s := 0
		for j := 0; j <= k; j++ {
			s += f[i-1][j]
			if j >= i {
				s -= f[i-1][j-i]
			}
			f[i][j] = s % mod
		}
	}
	return f[n][k]
}
